packages/sync/src/replicastore/class-table-checksum.php

```diff
@@ -0,0 +1,524 @@
+	 *
+	 * @param $table_configuration
+	 */
+	private function prepare_fields( $table_configuration ) {
+		$this->key_fields            = $table_configuration['key_fields'];
+		$this->range_field           = $table_configuration['range_field'];
+		$this->checksum_fields       = $table_configuration['checksum_fields'];
+		$this->additional_filter_sql = ! empty( $table_configuration['filter_sql'] ) ? $table_configuration['filter_sql'] : '';
+	}
+
+	/**
+	 * Verify provided table name is valid for checksum processing.
+	 *
+	 * @param $table
+	 *
+	 * @return mixed|string
+	 * @throws Exception
+	 */
+	private function validate_table_name( $table ) {
+		if ( empty( $table ) ) {
+			throw new Exception( 'Invalid table name: empty' );
+		}
+
+		if ( ! array_key_exists( $table, $this->allowed_tables ) ) {
+			throw new Exception( 'Invalid table name: not allowed' );
+		}
+
+		// TODO other checks if such are needed.
+
+		return $this->allowed_tables[ $table ]['table'];
+	}
+
+	/**
+	 * Verify provided fields are proper names.
+	 *
+	 * @param $fields
+	 *
+	 * @throws Exception
+	 */
+	private function validate_fields( $fields ) {
+		foreach ( $fields as $field ) {
+			if ( ! preg_match( '/^[0-9,a-z,A-Z$_]+$/i', $field ) ) {
+				throw new Exception( "Invalid field name: {$field} is not allowed" );
+			}
+
+			// TODO other verifications of the field names.
+		}
+	}
+
+	/**
+	 * Verify the fields exist in the table.
+	 *
+	 * @param $fields
+	 *
+	 * @return bool
+	 * @throws Exception
+	 */
+	private function validate_fields_against_table( $fields ) {
+		global $wpdb;
+
+		// TODO: Is this safe enough?
+		$result = $wpdb->get_row( "SELECT * FROM {$this->table} LIMIT 1", ARRAY_A );
+		if ( ! is_array( $result ) ) {
+			throw new Exception( 'Unexpected $wpdb->query output: not array' );
+		}
+
+		// Check if the fields are actually contained in the table.
+		foreach ( $fields as $field_to_check ) {
+			if ( ! array_key_exists( $field_to_check, $result ) ) {
+				throw new Exception( "Invalid field name: field '{$field_to_check}' doesn't exist in table {$this->table}" );
+			}
+		}
+
+		return true;
+	}
+
+	/**
+	 * Verify the configured fields.
+	 *
+	 * @throws Exception
+	 */
+	private function validate_input() {
+		$fields = array_merge( array( $this->range_field ), $this->key_fields, $this->checksum_fields );
+
+		$this->validate_fields( $fields );
+		$this->validate_fields_against_table( $fields );
+	}
+
+	/**
+	 * Build the filter query baased off range fields and values and the additional sql.
+	 *
+	 * @param null $range_from
+	 * @param null $range_to
+	 * @param null $filter_values
+	 *
+	 * @return string
+	 */
+	private function build_filter_statement( $range_from = null, $range_to = null, $filter_values = null ) {
+		global $wpdb;
+
+		/**
+		 * Prepare the ranges.
+		 */
+
+		$filter_array = array();
+		if ( $range_from !== null ) {
+			$filter_array[] = $wpdb->prepare( "{$this->range_field} > %d", array( intval( $range_from ) ) );
```

Comment:
Needs to be >= (inclusive)

Otherwise when we do buckets we miss the edge values.

 array_sum( $rs->checksum_histogram( 'posts', 2 ) );
=> 1933024224122
>>> array_sum( $rs->checksum_histogram( 'posts', 5 ) );
=> 1916103149414

[
     "min_range" => "1",
     "max_range" => "1278",
     "item_count" => "858",
   ]

Reply:
fixed in df1976b8f

Needed to have both >= and <= and during histograms we add +1 to the previous max_range value.